File: src/lib/trading/smc/structure.ts
// SMC Engine - Market Structure Detection Module
// Detects BOS (Break of Structure), CHoCH (Change of Character), HH, HL, LH, LL

import { Swing, SwingType, StructurePoint, StructureType, MarketDirection, Candle } from '../types';

export interface StructureConfig {
  minBreakPercent: number;  // Minimum percentage break to qualify
  confirmCandles: number;   // Candles needed for confirmation
}

const DEFAULT_CONFIG: StructureConfig = {
  minBreakPercent: 0.1,
  confirmCandles: 2
};

/**
 * Detect Market Structure from Swings
 * Identifies HH, HL, LH, LL patterns
 */
export function detectStructure(
  swings: Swing[],
  config: StructureConfig = DEFAULT_CONFIG
): StructurePoint[] {
  const structures: StructurePoint[] = [];

  if (swings.length < 2) {
    return structures;
  }

  // Sort swings by timestamp
  const sortedSwings = [...swings].sort((a, b) => 
    a.timestamp.getTime() - b.timestamp.getTime()
  );

  // Alternate between highs and lows
  let lastHigh: Swing | null = null;
  let lastLow: Swing | null = null;

  for (let i = 0; i < sortedSwings.length; i++) {
    const currentSwing = sortedSwings[i];
    const prevSwing = i > 0 ? sortedSwings[i - 1] : null;

    if (currentSwing.type === SwingType.HIGH) {
      if (lastHigh) {
        if (currentSwing.price > lastHigh.price) {
          // Higher High
          structures.push({
            symbolId: currentSwing.symbolId,
            timeframe: currentSwing.timeframe,
            timestamp: currentSwing.timestamp,
            type: StructureType.HH,
            direction: MarketDirection.BULLISH,
            price: currentSwing.price,
            swingId: currentSwing.id,
            confirmed: currentSwing.confirmed
          });
        } else {
          // Lower High
          structures.push({
            symbolId: currentSwing.symbolId,
            timeframe: currentSwing.timeframe,
            timestamp: currentSwing.timestamp,
            type: StructureType.LH,
            direction: MarketDirection.BEARISH,
            price: currentSwing.price,
            swingId: currentSwing.id,
            confirmed: currentSwing.confirmed
          });
        }
      }
      lastHigh = currentSwing;
    } else {
      if (lastLow) {
        if (currentSwing.price < lastLow.price) {
          // Lower Low
          structures.push({
            symbolId: currentSwing.symbolId,
            timeframe: currentSwing.timeframe,
            timestamp: currentSwing.timestamp,
            type: StructureType.LL,
            direction: MarketDirection.BEARISH,
            price: currentSwing.price,
            swingId: currentSwing.id,
            confirmed: currentSwing.confirmed
          });
        } else {
          // Higher Low
          structures.push({
            symbolId: currentSwing.symbolId,
            timeframe: currentSwing.timeframe,
            timestamp: currentSwing.timestamp,
            type: StructureType.HL,
            direction: MarketDirection.BULLISH,
            price: currentSwing.price,
            swingId: currentSwing.id,
            confirmed: currentSwing.confirmed
          });
        }
      }
      lastLow = currentSwing;
    }
  }

  return structures;
}

/**
 * Detect Break of Structure (BOS)
 * BOS occurs when price breaks a significant swing point
 */
export function detectBOS(
  candles: Candle[],
  swings: Swing[],
  config: StructureConfig = DEFAULT_CONFIG
): StructurePoint[] {
  const bosPoints: StructurePoint[] = [];

  if (candles.length < 5 || swings.length < 2) {
    return bosPoints;
  }

  // Get confirmed swing highs and lows
  const swingHighs = swings.filter(s => s.type === SwingType.HIGH && s.confirmed);
  const swingLows = swings.filter(s => s.type === SwingType.LOW && s.confirmed);

  // Detect bullish BOS (break above swing high)
  for (const swingHigh of swingHighs) {
    const swingIndex = candles.findIndex(c => 
      c.timestamp.getTime() === swingHigh.timestamp.getTime()
    );

    if (swingIndex === -1) continue;

    // Look for break after the swing
    for (let i = swingIndex + 1; i < candles.length; i++) {
      const candle = candles[i];
      const breakPercent = ((candle.close - swingHigh.price) / swingHigh.price) * 100;

      if (breakPercent >= config.minBreakPercent && candle.close > swingHigh.price) {
        // Check confirmation
        let confirmed = true;
        for (let j = i + 1; j < Math.min(i + 1 + config.confirmCandles, candles.length); j++) {
          if (candles[j].close < swingHigh.price) {
            confirmed = false;
            break;
          }
        }

        if (confirmed) {
          bosPoints.push({
            symbolId: candle.symbolId,
            timeframe: candle.timeframe,
            timestamp: candle.timestamp,
            type: StructureType.BOS,
            direction: MarketDirection.BULLISH,
            price: candle.close,
            brokenLevel: swingHigh.price,
            confirmed: true
          });
          break;
        }
      }
    }
  }

  // Detect bearish BOS (break below swing low)
  for (const swingLow of swingLows) {
    const swingIndex = candles.findIndex(c => 
      c.timestamp.getTime() === swingLow.timestamp.getTime()
    );

    if (swingIndex === -1) continue;

    for (let i = swingIndex + 1; i < candles.length; i++) {
      const candle = candles[i];
      const breakPercent = ((swingLow.price - candle.close) / swingLow.price) * 100;

      if (breakPercent >= config.minBreakPercent && candle.close < swingLow.price) {
        let confirmed = true;
        for (let j = i + 1; j < Math.min(i + 1 + config.confirmCandles, candles.length); j++) {
          if (candles[j].close > swingLow.price) {
            confirmed = false;
            break;
          }
        }

        if (confirmed) {
          bosPoints.push({
            symbolId: candle.symbolId,
            timeframe: candle.timeframe,
            timestamp: candle.timestamp,
            type: StructureType.BOS,
            direction: MarketDirection.BEARISH,
            price: candle.close,
            brokenLevel: swingLow.price,
            confirmed: true
          });
          break;
        }
      }
    }
  }

  return bosPoints;
}

/**
 * Detect Change of Character (CHoCH)
 * CHoCH indicates a potential trend reversal
 */
export function detectCHoCH(
  candles: Candle[],
  swings: Swing[],
  structures: StructurePoint[]
): StructurePoint[] {
  const chochPoints: StructurePoint[] = [];

  if (structures.length < 3) {
    return chochPoints;
  }

  // Sort structures by timestamp
  const sortedStructures = [...structures].sort((a, b) => 
    a.timestamp.getTime() - b.timestamp.getTime()
  );

  // Look for CHoCH patterns
  for (let i = 2; i < sortedStructures.length; i++) {
    const current = sortedStructures[i];
    const prev1 = sortedStructures[i - 1];
    const prev2 = sortedStructures[i - 2];

    // Bullish CHoCH: After series of LLs and LHs, we get HH
    if (
      prev2.type === StructureType.LL &&
      prev1.type === StructureType.LH &&
      current.type === StructureType.HH
    ) {
      chochPoints.push({
        symbolId: current.symbolId,
        timeframe: current.timeframe,
        timestamp: current.timestamp,
        type: StructureType.CHOCH,
        direction: MarketDirection.BULLISH,
        price: current.price,
        confirmed: current.confirmed
      });
    }

    // Bearish CHoCH: After series of HHs and HLs, we get LL
    if (
      prev2.type === StructureType.HH &&
      prev1.type === StructureType.HL &&
      current.type === StructureType.LL
    ) {
      chochPoints.push({
        symbolId: current.symbolId,
        timeframe: current.timeframe,
        timestamp: current.timestamp,
        type: StructureType.CHOCH,
        direction: MarketDirection.BEARISH,
        price: current.price,
        confirmed: current.confirmed
      });
    }
  }

  return chochPoints;
}

/**
 * Determine current market trend direction
 */
export function getTrendDirection(structures: StructurePoint[]): MarketDirection {
  if (structures.length < 2) {
    return MarketDirection.NEUTRAL;
  }

  const recentStructures = structures.slice(-4);
  let bullishCount = 0;
  let bearishCount = 0;

  for (const s of recentStructures) {
    if (s.direction === MarketDirection.BULLISH) {
      bullishCount++;
    } else if (s.direction === MarketDirection.BEARISH) {
      bearishCount++;
    }
  }

  if (bullishCount > bearishCount) {
    return MarketDirection.BULLISH;
  } else if (bearishCount > bullishCount) {
    return MarketDirection.BEARISH;
  }

  return MarketDirection.NEUTRAL;
}

/**
 * Check if structure is broken
 */
export function isStructureBroken(
  candles: Candle[],
  structure: StructurePoint
): boolean {
  const structIndex = candles.findIndex(c => 
    c.timestamp.getTime() === structure.timestamp.getTime()
  );

  if (structIndex === -1) return false;

  for (let i = structIndex + 1; i < candles.length; i++) {
    if (structure.direction === MarketDirection.BULLISH) {
      if (candles[i].close < structure.price) {
        return true;
      }
    } else {
      if (candles[i].close > structure.price) {
        return true;
      }
    }
  }

  return false;
}

File: mini-services/trading-engine/app/smc/structure.py
"""
SMC Engine - Market Structure Detection Module
Mathematical implementation of BOS, CHoCH, HH, HL, LH, LL

Structure Classification:
- HH (Higher High): new_high > previous_high
- HL (Higher Low): new_low > previous_low
- LH (Lower High): new_high < previous_high
- LL (Lower Low): new_low < previous_low

BOS (Break of Structure):
- Bullish: close > last_swing_high
- Bearish: close < last_swing_low

CHoCH (Change of Character):
- Trend reversal signal
- Uptrend → LL break
- Downtrend → HH break

Author: Trading AI Agent
"""

from dataclasses import dataclass
from typing import List, Optional, Tuple
from enum import Enum
from datetime import datetime
import numpy as np

from .swing import Swing, Candle


class StructureType(Enum):
    """Market Structure Types"""
    HH = "HH"  # Higher High
    HL = "HL"  # Higher Low
    LH = "LH"  # Lower High
    LL = "LL"  # Lower Low
    BOS = "BOS"  # Break of Structure
    CHOCH = "CHOCH"  # Change of Character


class TrendDirection(Enum):
    """Trend Direction"""
    BULLISH = "BULLISH"
    BEARISH = "BEARISH"
    NEUTRAL = "NEUTRAL"


@dataclass
class StructurePoint:
    """Market Structure Point"""
    timestamp: datetime
    type: StructureType
    direction: TrendDirection
    price: float
    swing_id: Optional[int] = None
    broken_level: Optional[float] = None
    confirmed: bool = False
    candle_index: int = 0


class StructureDetector:
    """
    Market Structure Detection using mathematical formulas
    
    Identifies:
    - Higher Highs (HH), Higher Lows (HL) - Bullish structure
    - Lower Highs (LH), Lower Lows (LL) - Bearish structure
    - Break of Structure (BOS)
    - Change of Character (CHoCH)
    """
    
    def __init__(self, min_break_percent: float = 0.1, confirm_candles: int = 2):
        """
        Initialize Structure Detector
        
        Args:
            min_break_percent: Minimum percentage break to qualify as BOS
            confirm_candles: Candles needed for confirmation
        """
        self.min_break_percent = min_break_percent
        self.confirm_candles = confirm_candles
    
    def detect_basic_structure(self, swings: List[Swing]) -> List[StructurePoint]:
        """
        Detect basic structure (HH, HL, LH, LL) from swings
        
        Mathematical formulas:
        - HH: new_high > previous_high
        - LL: new_low < previous_low
        - HL: new_low > previous_low (in uptrend)
        - LH: new_high < previous_high (in downtrend)
        """
        if len(swings) < 2:
            return []
        
        structures = []
        
        # Sort swings by timestamp
        sorted_swings = sorted(swings, key=lambda s: s.timestamp)
        
        last_high: Optional[Swing] = None
        last_low: Optional[Swing] = None
        
        for swing in sorted_swings:
            if swing.type == 'HIGH':
                if last_high is not None:
                    if swing.price > last_high.price:
                        # Higher High - Bullish
                        structures.append(StructurePoint(
                            timestamp=swing.timestamp,
                            type=StructureType.HH,
                            direction=TrendDirection.BULLISH,
                            price=swing.price,
                            swing_id=id(swing),
                            confirmed=swing.confirmed,
                            candle_index=swing.candle_index
                        ))
                    else:
                        # Lower High - Bearish
                        structures.append(StructurePoint(
                            timestamp=swing.timestamp,
                            type=StructureType.LH,
                            direction=TrendDirection.BEARISH,
                            price=swing.price,
                            swing_id=id(swing),
                            confirmed=swing.confirmed,
                            candle_index=swing.candle_index
                        ))
                last_high = swing
            
            else:  # LOW
                if last_low is not None:
                    if swing.price < last_low.price:
                        # Lower Low - Bearish
                        structures.append(StructurePoint(
                            timestamp=swing.timestamp,
                            type=StructureType.LL,
                            direction=TrendDirection.BEARISH,
                            price=swing.price,
                            swing_id=id(swing),
                            confirmed=swing.confirmed,
                            candle_index=swing.candle_index
                        ))
                    else:
                        # Higher Low - Bullish
                        structures.append(StructurePoint(
                            timestamp=swing.timestamp,
                            type=StructureType.HL,
                            direction=TrendDirection.BULLISH,
                            price=swing.price,
                            swing_id=id(swing),
                            confirmed=swing.confirmed,
                            candle_index=swing.candle_index
                        ))
                last_low = swing
        
        return structures
    
    def detect_bos(self, candles: List[Candle], swings: List[Swing]) -> List[StructurePoint]:
        """
        Detect Break of Structure (BOS)
        
        Mathematical condition:
        - Bullish BOS: close > last_swing_high
        - Bearish BOS: close < last_swing_low
        
        With confirmation:
        - Must close beyond the level with strong candle
        """
        if len(candles) < 5 or len(swings) < 2:
            return []
        
        bos_points = []
        
        # Get confirmed swing highs and lows
        swing_highs = sorted([s for s in swings if s.type == 'HIGH' and s.confirmed],
                            key=lambda s: s.candle_index)
        swing_lows = sorted([s for s in swings if s.type == 'LOW' and s.confirmed],
                           key=lambda s: s.candle_index)
        
        # Detect Bullish BOS (break above swing high)
        for swing_high in swing_highs:
            swing_idx = swing_high.candle_index
            
            # Look for break after the swing
            for i in range(swing_idx + 1, len(candles)):
                candle = candles[i]
                
                # Mathematical condition: close > swing_high.price
                break_percent = ((candle.close - swing_high.price) / swing_high.price) * 100
                
                if break_percent >= self.min_break_percent and candle.close > swing_high.price:
                    # Check confirmation (closes stay above)
                    confirmed = True
                    for j in range(i + 1, min(i + 1 + self.confirm_candles, len(candles))):
                        if candles[j].close < swing_high.price:
                            confirmed = False
                            break
                    
                    if confirmed:
                        bos_points.append(StructurePoint(
                            timestamp=candle.timestamp,
                            type=StructureType.BOS,
                            direction=TrendDirection.BULLISH,
                            price=candle.close,
                            broken_level=swing_high.price,
                            confirmed=True,
                            candle_index=i
                        ))
                        break  # Only one BOS per swing level
        
        # Detect Bearish BOS (break below swing low)
        for swing_low in swing_lows:
            swing_idx = swing_low.candle_index
            
            for i in range(swing_idx + 1, len(candles)):
                candle = candles[i]
                
                # Mathematical condition: close < swing_low.price
                break_percent = ((swing_low.price - candle.close) / swing_low.price) * 100
                
                if break_percent >= self.min_break_percent and candle.close < swing_low.price:
                    confirmed = True
                    for j in range(i + 1, min(i + 1 + self.confirm_candles, len(candles))):
                        if candles[j].close > swing_low.price:
                            confirmed = False
                            break
                    
                    if confirmed:
                        bos_points.append(StructurePoint(
                            timestamp=candle.timestamp,
                            type=StructureType.BOS,
                            direction=TrendDirection.BEARISH,
                            price=candle.close,
                            broken_level=swing_low.price,
                            confirmed=True,
                            candle_index=i
                        ))
                        break
        
        return bos_points
    
    def detect_choch(self, structures: List[StructurePoint]) -> List[StructurePoint]:
        """
        Detect Change of Character (CHoCH)
        
        Mathematical pattern:
        - Bullish CHoCH: After LL + LH, we get HH (downtrend to uptrend)
        - Bearish CHoCH: After HH + HL, we get LL (uptrend to downtrend)
        """
        if len(structures) < 3:
            return []
        
        choch_points = []
        
        # Sort by candle index
        sorted_structures = sorted(structures, key=lambda s: s.candle_index)
        
        for i in range(2, len(sorted_structures)):
            prev2 = sorted_structures[i - 2]
            prev1 = sorted_structures[i - 1]
            current = sorted_structures[i]
            
            # Bullish CHoCH: LL + LH + HH
            if (prev2.type == StructureType.LL and
                prev1.type == StructureType.LH and
                current.type == StructureType.HH):
                choch_points.append(StructurePoint(
                    timestamp=current.timestamp,
                    type=StructureType.CHOCH,
                    direction=TrendDirection.BULLISH,
                    price=current.price,
                    confirmed=current.confirmed,
                    candle_index=current.candle_index
                ))
            
            # Bearish CHoCH: HH + HL + LL
            elif (prev2.type == StructureType.HH and
                  prev1.type == StructureType.HL and
                  current.type == StructureType.LL):
                choch_points.append(StructurePoint(
                    timestamp=current.timestamp,
                    type=StructureType.CHOCH,
                    direction=TrendDirection.BEARISH,
                    price=current.price,
                    confirmed=current.confirmed,
                    candle_index=current.candle_index
                ))
        
        return choch_points
    
    def get_trend_direction(self, structures: List[StructurePoint]) -> TrendDirection:
        """
        Determine current market trend direction
        
        Analyzes recent structures to determine bias
        """
        if len(structures) < 2:
            return TrendDirection.NEUTRAL
        
        # Get recent structures
        recent = structures[-4:] if len(structures) >= 4 else structures
        
        bullish_count = sum(1 for s in recent if s.direction == TrendDirection.BULLISH)
        bearish_count = sum(1 for s in recent if s.direction == TrendDirection.BEARISH)
        
        if bullish_count > bearish_count:
            return TrendDirection.BULLISH
        elif bearish_count > bullish_count:
            return TrendDirection.BEARISH
        
        return TrendDirection.NEUTRAL


def analyze_structure(candles: List[Candle], swings: List[Swing]) -> dict:
    """
    Complete structure analysis
    
    Returns dict with:
    - basic_structures: HH, HL, LH, LL
    - bos_points: Break of Structure
    - choch_points: Change of Character
    - trend: Current trend direction
    """
    detector = StructureDetector()
    
    basic = detector.detect_basic_structure(swings)
    bos = detector.detect_bos(candles, swings)
    choch = detector.detect_choch(basic)
    trend = detector.get_trend_direction(basic)
    
    return {
        'structures': basic,
        'bos_points': bos,
        'choch_points': choch,
        'trend': trend.value,
        'all_structures': sorted(basic + bos + choch, key=lambda s: s.candle_index)
    }
